### ISSUE TITLE
Automatic differentiation

### ISSUE THREAD

**Issue by jcallaham** [dynamicslab/hydrogym#8]
This will be an alternative to black-box training schemes like RL.  There are (at least) three cases we should support:
1. Parameter optimization within Firedrake (i.e. standard optimal control)
2. `hydrogym.Flow` as a `torch.nn.Module` so that parameters can be optimized within PyTorch
3. Embedding a `torch.nn.Module` as an `Expression` (or similar) within Firedrake

The last two cases are sort of opposite.  Basically, you should be able to either put the CFD inside an ML model and differentiate with PyTorch, or have an ML model within the CFD code and differentiate with pyadjoint.  This will probably be a matter of figuring out how to pass information back and forth between the gradient tapes.  Compatibility with other libraries could be nice, but this would be a good place to start.

### Goals:
- [x] Implement automatic differentiation with pyadjoint
- [x] Profile memory usage: is checkpointing necessary?
- [x] #14
- [x] #15
